File: lib/measures/improve_fan_belt_efficiency/measure.rb
# frozen_string_literal: true

# *******************************************************************************
# open_studio(R), Copyright (c) Alliance for Sustainable Energy, llc.
# See also https://openstudio.net/license
# *******************************************************************************

# start the measure
class temp_class_0 < open_studio::Measure::model_measure
  # define the name that a user will see, this method may be deprecated as
  # the display name in pat comes from the name field in measure.xml
  def name
    return 'Improve Fan Belt Efficiency'
  end

  # define the arguments that the user will input
  def arguments(model)
    args = open_studio::Measure::os_argument_vector.new

    # populate choice argument for constructions that are applied to surfaces in the model
    loop_handles = open_studio::String_vector.new
    loop_display_names = open_studio::String_vector.new

    # putting air loops and names into hash
    loop_args = model.get_air_loop_hva_cs
    loop_args_hash = {}
    loop_args.each do |loop_arg|
      loop_args_hash[loop_arg.name.to_s] = loop_arg
    end

    # looping through sorted hash of air loops
    loop_args_hash.sort.map do |key, value|
      show_loop = false
      components = value.supply_components
      components.each do |component|
        if !component.to_fan_constant_volume.empty?
          show_loop = true
        end
        if !component.to_fan_variable_volume.empty?
          show_loop = true
        end
        if !component.to_fan_on_off.empty?
          show_loop = true
        end
      end

      # if loop as object of correct type then add to hash.
      if show_loop == true
        loop_handles << value.handle.to_s
        loop_display_names << key
      end
    end

    # add building to string vector with air loops
    building = model.get_building
    loop_handles << building.handle.to_s
    loop_display_names << '*All Air Loops*'

    # make an argument for air loops
    object = open_studio::Measure::os_argument.make_choice_argument('object', loop_handles, loop_display_names, true)
    object.set_display_name('Choose an Air Loop to Alter.')
    object.set_default_value('*All Air Loops*') # if no loop is chosen this will run on all air loops
    args << object

    # todo: - change this to choice list from design document
    # make an argument to add new space true/false
    motor_eff = open_studio::Measure::os_argument.make_double_argument('motor_eff', true)
    motor_eff.set_display_name('Motor Efficiency Improvement Due to Fan Belt Improvements(%).')
    motor_eff.set_default_value(3.0)
    args << motor_eff

    # bool argument to remove existing costs
    remove_costs = open_studio::Measure::os_argument.make_bool_argument('remove_costs', true)
    remove_costs.set_display_name('Remove Baseline Costs From Effected Fans?')
    remove_costs.set_default_value(false)
    args << remove_costs

    # make an argument for material and installation cost
    material_cost = open_studio::Measure::os_argument.make_double_argument('material_cost', true)
    material_cost.set_display_name('Material and Installation Costs per Motor ($).')
    material_cost.set_default_value(0.0)
    args << material_cost

    # make an argument for demolition cost
    demolition_cost = open_studio::Measure::os_argument.make_double_argument('demolition_cost', true)
    demolition_cost.set_display_name('Demolition Costs per Motor ($).')
    demolition_cost.set_default_value(0.0)
    args << demolition_cost

    # make an argument for duration in years until costs start
    years_until_costs_start = open_studio::Measure::os_argument.make_integer_argument('years_until_costs_start', true)
    years_until_costs_start.set_display_name('Years Until Costs Start (whole years).')
    years_until_costs_start.set_default_value(0)
    args << years_until_costs_start

    # make an argument to determine if demolition costs should be included in initial construction
    demo_cost_initial_const = open_studio::Measure::os_argument.make_bool_argument('demo_cost_initial_const', true)
    demo_cost_initial_const.set_display_name('Demolition Costs Occur During Initial Construction?')
    demo_cost_initial_const.set_default_value(false)
    args << demo_cost_initial_const

    # make an argument for expected life
    expected_life = open_studio::Measure::os_argument.make_integer_argument('expected_life', true)
    expected_life.set_display_name('Expected Life (whole years).')
    expected_life.set_default_value(20)
    args << expected_life

    # make an argument for o&m cost
    om_cost = open_studio::Measure::os_argument.make_double_argument('om_cost', true)
    om_cost.set_display_name('O & M Costs per Motor ($).')
    om_cost.set_default_value(0.0)
    args << om_cost

    # make an argument for o&m frequency
    om_frequency = open_studio::Measure::os_argument.make_integer_argument('om_frequency', true)
    om_frequency.set_display_name('O & M Frequency (whole years).')
    om_frequency.set_default_value(1)
    args << om_frequency

    return args
  end

  # define what happens when the measure is cop
  def run(model, runner, user_arguments)
    super(model, runner, user_arguments)

    # use the built-in error checking
    if !runner.validate_user_arguments(arguments(model), user_arguments)
      return false
    end

    # assign the user inputs to variables
    object = runner.get_optional_workspace_object_choice_value('object', user_arguments, model) # model is passed in because of argument type
    motor_eff = runner.get_double_argument_value('motor_eff', user_arguments)
    remove_costs = runner.get_bool_argument_value('remove_costs', user_arguments)
    material_cost = runner.get_double_argument_value('material_cost', user_arguments)
    demolition_cost = runner.get_double_argument_value('demolition_cost', user_arguments)
    years_until_costs_start = runner.get_integer_argument_value('years_until_costs_start', user_arguments)
    demo_cost_initial_const = runner.get_bool_argument_value('demo_cost_initial_const', user_arguments)
    expected_life = runner.get_integer_argument_value('expected_life', user_arguments)
    om_cost = runner.get_double_argument_value('om_cost', user_arguments)
    om_frequency = runner.get_integer_argument_value('om_frequency', user_arguments)

    # check the loop for reasonableness
    apply_to_all_loops = false
    loop = nil
    if object.empty?
      handle = runner.get_string_argument_value('loop', user_arguments)
      if handle.empty?
        runner.register_error('No loop was chosen.')
      else
        runner.register_error("The selected loop with handle '#{handle}' was not found in the model. It may have been removed by another measure.")
      end
      return false
    else
      if !object.get.to_loop.empty?
        loop = object.get.to_loop.get
      elsif !object.get.to_building.empty?
        apply_to_all_loops = true
      else
        runner.register_error('Script Error - argument not showing up as loop.')
        return false
      end
    end

    # check the user_name for reasonableness
    if (motor_eff <= 1) ||  (motor_eff >= 5)
      runner.register_warning('Requested motor efficiency improvement is not between expected values of 1% and 5%')
    end
    # motor efficiency will be checked motor by motor to see warn if higher than 0.96 and error if not between or equal to 0 and 1

    # set flags to use later
    costs_requested = false

    # set values to use later
    yr0_capital_total_costs_baseline = 0
    yr0_capital_total_costs_proposed = 0

    # If demo_cost_initial_const is true then will be applied once in the lifecycle. Future replacements use the demo cost of the new construction.
    demo_costs_of_baseline_objectss = 0

    # check costs for reasonableness
    if material_cost.abs + demolition_cost.abs + om_cost.abs == 0
      runner.register_info('No costs were requested for motors improvements.')
    else
      costs_requested = true
    end

    # check lifecycle arguments for reasonableness
    if (years_until_costs_start < 0) && (years_until_costs_start > expected_life)
      runner.register_error('Years until costs start should be a non-negative integer less than Expected Life.')
    end
    if (expected_life < 1) && (expected_life > 100)
      runner.register_error('Choose an integer greater than 0 and less than or equal to 100 for Expected Life.')
    end
    if om_frequency < 1
      runner.register_error('Choose an integer greater than 0 for O & M Frequency.')
    end

    # short def to make numbers pretty (converts 4125001.25641 to 4,125,001.26 or 4,125,001). The definition be called through this measure
    # round to 0 or 2)
    neat_numbers = lambda do |number, roundto = 2|
            number.to_s.reverse.gsub(/([0-9]{3}(?=([0-9])))/, '\\1,').reverse
    end

    # helper that loops through lifecycle costs getting total costs under "Construction" or "Salvage" category and add to counter if occurs during year 0
    get_total_costs_for_objects = lambda do |objects|
                  runner.register_warning("Requested efficiency of #{target_motor_efficiency * 100}% for #{h_vac_component.name} is not possible. Setting motor efficiency to 100%.")
                elsif target_motor_efficiency < 0
                  h_vac_component.set_motor_efficiency(0.0)
                  runner.register_warning("Requested efficiency of #{target_motor_efficiency * 100}% for #{h_vac_component.name} is not possible. Setting motor efficiency to 0%.")
                else
                  h_vac_component.set_motor_efficiency(target_motor_efficiency)
                  runner.register_info("Changing the motor efficiency from #{initial_motor_efficiency * 100}% to #{target_motor_efficiency * 100}% for '#{h_vac_component.name}' onloop '#{loop.name}.'")
                  if target_motor_efficiency > 0.96
                    runner.register_warning("Requested efficiency for #{h_vac_component.name} is greater than 96%.")
                  end
                end
      
                # get initial year 0 cost
                yr0_capital_total_costs_baseline += get_total_costs_for_objects([h_vac_component])
      
                # demo value of baseline costs associated with unit
                demo_lc_cs = h_vac_component.life_cycle_costs
                demo_lc_cs.each do |demo_lcc|
                  if demo_lcc.category == 'Salvage'
                    demo_costs_of_baseline_objectss += demo_lcc.total_cost
                  end
                end
      
                # remove all old costs
                if !h_vac_component.life_cycle_costs.empty? && (remove_costs == true)
                  runner.register_info("Removing existing lifecycle cost objects associated with #{h_vac_component.name}")
                  removed_costs = h_vac_component.remove_life_cycle_costs
                end
      
                # add new costs
                if costs_requested == true
      
                  # adding new cost items
                  lcc_mat = open_studio::Model::Life_cycle_cost.create_life_cycle_cost("lcc_mat - #{h_vac_component.name}", h_vac_component, material_cost, 'cost_per_each', 'Construction', expected_life, years_until_costs_start)
                  # cost for if demo_initial_construction == true is added at the end of the measure
                  lcc_demo = open_studio::Model::Life_cycle_cost.create_life_cycle_cost("lcc_demo - #{h_vac_component.name}", h_vac_component, demolition_cost, 'cost_per_each', 'Salvage', expected_life, years_until_costs_start + expected_life)
                  lcc_om = open_studio::Model::Life_cycle_cost.create_life_cycle_cost("lcc_om - #{h_vac_component.name}", h_vac_component, om_cost, 'cost_per_each', 'Maintenance', om_frequency, 0)
      
                  # get final year 0 cost
                  yr0_capital_total_costs_proposed += get_total_costs_for_objects([h_vac_component])
      
                end
      
              end
            end
    end

    # add one time demo cost of removed windows if appropriate
    if demo_cost_initial_const == true
      building = model.get_building
      lcc_baseline_demo = open_studio::Model::Life_cycle_cost.create_life_cycle_cost('lcc_baseline_demo', building, demo_costs_of_baseline_objectss, 'cost_per_each', 'Salvage', 0, years_until_costs_start).get # using 0 for repeat period since one time cost.
      runner.register_info("Adding one time cost of $#{neat_numbers(lcc_baseline_demo.total_cost, 0)} related to demolition of baseline objects.")

      # if demo occurs on year 0 then add to initial capital cost counter
      if lcc_baseline_demo.years_from_start == 0
        yr0_capital_total_costs_proposed += lcc_baseline_demo.total_cost
      end
    end

    if initial_motor_efficiency_values.size + missing_initial_motor_efficiency == 0
      runner.register_as_not_applicable('The affected loop(s) does not contain any fans, the model will not be altered.')
      return true
    end

    # reporting initial condition of model
    runner.register_initial_condition("The starting motor efficiency values in affected loop(s) range from #{initial_motor_efficiency_values.min * 100}% to #{initial_motor_efficiency_values.max * 100}%. Initial year 0 capital costs for affected fans is $#{neat_numbers(yr0_capital_total_costs_baseline, 0)}.")

    # reporting final condition of model
    runner.register_final_condition("#{initial_motor_efficiency_values.size + missing_initial_motor_efficiency} fans had motor efficiency values set to altered. Final year 0 capital costs for affected fans is $#{neat_numbers(yr0_capital_total_costs_proposed, 0)}.")

    return true
  end
end

# this allows the measure to be used by the application
Improve_fan_belt_efficiency.new.register_with_application

File: lib/measures/ReplaceExteriorWindowConstruction/measure.py
# *******************************************************************************
# OpenStudio(R), Copyright (c) Alliance for Sustainable Energy, LLC.
# See also https://openstudio.net/license
# *******************************************************************************
import openstudio
import typing
import re
from pint import UnitRegistry

# start the measure
class ReplaceExteriorWindowConstruction(openstudio.measure.ModelMeasure):
  # define the name that a user will see
  def name(self):
    return "Replace Exterior Window Constructions with a Different Construction from the Model."

  # human readable description
  def description(self):
    return "Replace existing windows with different windows to change thermal or lighting performance.  Window technology has improved drastically over the years, and double or triple-pane high performance windows currently on the market can cut down on envelope loads greatly.  Window frames with thermal breaks reduce the considerable energy that can transfer through thermally unbroken frames.  High performance windows typically also come with low-emissivity (low?e) glass to keep radiant heat on the same side of the glass from where the heat originated. This means that during the cooling months a low-e glass would tend to keep radiant heat from the sun on the outside of the window, which would keep the inside of a building cooler. Conversely, during heating months low-e glass helps keep radiant heat from inside the building on the inside, which would keep the inside of a building warmer.  Life cycle cost values may be added for the new window applied by the measure."

  # human readable description of modeling approach
  def modeler_description(self):
    return "Replace fixed and/or operable exterior window constructions with another construction in the model.  Skylights (windows in roofs vs. walls) will not be altered. Windows in surfaces with `Adiabatic` boundary conditions are not specifically assumed to be interior or exterior. As a result constructions used on windows in `Adiabatic` surfaces will not be altered. `Material, installation, demolition, and O and M costs` can be added to the applied window construction. Optionally any prior costs associated with construction can be removed. <br/> <br/> For costs added as part of a design alternatives `Years Until Costs Start?` is typically `0`. For a new construction scenario `Demolition Costs Occur During Initial Construction?` is `false`. For retrofit scenario `Demolition Costs Occur During Initial Construction?` is `true`. `O and M cost and frequency` can be whatever is appropriate for the component."

  # define the arguments that the user will input
  def arguments(self, model: typing.Optional[openstudio.model.Model] = None):
    args = openstudio.measure.OSArgumentVector()

    # make a choice argument for constructions that are appropriate for windows
    construction_handles = openstudio.makeStringVector.new()
    construction_display_names = openstudio.makeStringVector.new()

    # putting space types and names into hash
    construction_args = model.getConstructions(model)
    construction_args_hash = {}
    for construction_arg in construction_args:
      construction_args_hash[construction_arg.name] = construction_arg

    # looping through sorted hash of constructions
    for key, value in sorted(construction_args_hash.items()):
    # Only include if construction is a valid fenestration construction
      if value.isFenestration:
        construction_handles.append(str(value.handle))
        construction_display_names.append(key)

    # make a choice argument for fixed windows
    construction = openstudio.measure.OSArgument.makeChoiceArgument("construction", construction_handles, construction_display_names, True)
    construction.setDisplayName("Pick a Window Construction From the Model to Replace Existing Window Constructions.")
    args.append(construction)

    # make a bool argument for fixed windows
    change_fixed_windows = openstudio.measure.OSArgument.makeBoolArgument("change_fixed_windows", True)
    change_fixed_windows.setDisplayName("Change Fixed Windows?")
    change_fixed_windows.setDefaultValue(True)
    args.append(change_fixed_windows)

    # make a bool argument for operable windows
    change_operable_windows = openstudio.measure.OSArgument.makeBoolArgument("change_operable_windows", True)
    change_operable_windows.setDisplayName("Change Operable Windows?")
    change_operable_windows.setDefaultValue(True)
    args.append(change_operable_windows)

    # make an argument to remove existing costs
    remove_costs = openstudio.measure.OSArgument.makeBoolArgument("remove_costs", True)
    remove_costs.setDisplayName("Remove Existing Costs?")
    remove_costs.setDefaultValue(True)
    args.append(remove_costs)

    # make an argument for material and installation cost
    material_cost_ip = openstudio.measure.OSArgument.makeDoubleArgument("material_cost_ip", True)
    material_cost_ip.setDisplayName("Material and Installation Costs for Construction per Area Used ($/ft^2).")
    material_cost_ip.setDefaultValue(0.0)
    args.append(material_cost_ip)

    # make an argument for demolition cost
    demolition_cost_ip = openstudio.measure.OSArgument.makeDoubleArgument("demolition_cost_ip", True)
    demolition_cost_ip.setDisplayName("Demolition Costs for Construction per Area Used ($/ft^2).")
    demolition_cost_ip.setDefaultValue(0.0)
    args.append(demolition_cost_ip)

    # make an argument for duration in years until costs start
    years_until_costs_start = openstudio.measure.OSArgument.makeIntegerArgument("years_until_costs_start", True)
    years_until_costs_start.setDisplayName("Years Until Costs Start (whole years).")
    years_until_costs_start.setDefaultValue(0)
    args.append(years_until_costs_start)

    # make an argument to determine if demolition costs should be included in initial construction
    demo_cost_initial_const = openstudio.measure.OSArgument.makeBoolArgument("demo_cost_initial_const", True)
    demo_cost_initial_const.setDisplayName("Demolition Costs Occur During Initial Construction?")
    demo_cost_initial_const.setDefaultValue(False)
    args.append(demo_cost_initial_const)

    # make an argument for expected life
    expected_life = openstudio.measure.OSArgument.makeIntegerArgument('expected_life', True)
    expected_life.setDisplayName("Expected Life (whole years).")
    expected_life.setDefaultValue(20)
    args.append(expected_life)

    # make an argument for o&m cost
    om_cost_ip = openstudio.measure.OSArgument.makeDoubleArgument("om_cost_ip", True)
    om_cost_ip.setDisplayName("O & M Costs for Construction per Area Used ($/ft^2).")
    om_cost_ip.setDefaultValue(0.0)
    args.append(om_cost_ip)

    # make an argument for o&m frequency
    om_frequency = openstudio.measure.OSArgument.makeIntegerArgument("om_frequency", True)
    om_frequency.setDisplayName("O & M Frequency (whole years).")
    om_frequency.setDefaultValue(1)
    args.append(om_frequency)

    return args

  # define what happens when the measure is run
  def run(self, model: openstudio.model.Model, runner: openstudio.measure.OSRunner, user_arguments: openstudio.measure.OSArgumentMap):
    """Execute the measure."""
    super().run(model, runner, user_arguments)  # Required

    # use the built-in error checking
    if not runner.validateUserArguments(self.arguments(model), user_arguments):
      return False

    # assign the user inputs to variables
    construction = runner.getOptionalWorkspaceObjectChoiceValue("construction", user_arguments, model)
    change_fixed_windows = runner.getBoolArgumentValue("change_fixed_windows", user_arguments)
    change_operable_windows = runner.getBoolArgumentValue("change_operable_windows", user_arguments)
    remove_costs = runner.getBoolArgumentValue("remove_costs", user_arguments)
    material_cost_ip = runner.getDoubleArgumentValue("material_cost_ip", user_arguments)
    demolition_cost_ip = runner.getDoubleArgumentValue("demolition_cost_ip", user_arguments)
    years_until_costs_start = runner.getIntegerArgumentValue("years_until_costs_start", user_arguments)
    demo_cost_initial_const = runner.getBoolArgumentValue("demo_cost_initial_const", user_arguments)
    expected_life = runner.getIntegerArgumentValue("expected_life", user_arguments)
    om_cost_ip = runner.getDoubleArgumentValue("om_cost_ip", user_arguments)
    om_frequency = runner.getIntegerArgumentValue("om_frequency", user_arguments)

    # check the construction for reasonableness
    if not construction:
      handle = runner.getStringArgumentValue("construction", user_arguments)
      if not handle:
        runner.registerError('No construction was chosen.')
      else:
        runner.registerError(
            f"The selected construction with handle '{handle}' was not found in the model. "
            "It may have been removed by another measure."
        )
      return False
    else:
      if construction.get().to_Construction(): # Assuming get() returns a valid object
        construction = construction.get.to_Construction().get()
      else:
        runner.registerError("Script Error - argument not showing up as construction.")
        return False

    # set flags and counters to use later
    costs_requested = False
    costs_removed = False

    # Later will add hard sized $ cost to this each time I swap a construction surfaces.
    # If demo_cost_initial_const is true then will be applied once in the lifecycle. Future replacements use the demo cost of the new construction.
    demo_costs_of_baseline_objects = 0

    # check costs for reasonableness
    if abs(material_cost_ip) + abs(demolition_cost_ip) + abs(om_cost_ip) == 0:
      runner.registerInfo(f"No costs were requested for {construction.name}.")
    else:
      costs_requested = True

    # check lifecycle arguments for reasonableness
    if years_until_costs_start < 0 or years_until_costs_start > expected_life:
      runner.registerError("Years until costs start should be a non-negative integer less than Expected Life.")

    if expected_life < 1 or expected_life > 100:
      runner.registerError("Choose an integer greater than 0 and less than or equal to 100 for Expected Life.")

    if om_frequency < 1:
      runner.registerError("Choose an integer greater than 0 for O & M Frequency.")

    # short def to make numbers pretty (converts 4125001.25641 to 4,125,001.26 or 4,125,001). The definition be called through this measure
    def neat_numbers(number, roundto = 2): # round to 0 or 2
      if roundto == 2:
        number = f"{number:.2f}"
      else:
        number = round(number)
      
      # Regex to add commas
      return re.sub(r'(?<=\d)(?=(\d{3})+(?!\d))', ',', str(number))

    # clone construction to get proper area for measure economics, in case it is used elsewhere in the building
    new_object = construction.clone(model)
    if new_object.to_Construction(): # Assuming it returns a valid object
      construction = new_object.to_Construction().get()

    # remove any component cost line items associated with the construction.
    if construction.lifeCycleCosts() & remove_costs == True:
      runner.registerInfo(f"Removing existing lifecycle cost objects associated with {construction.name}")
      removed_costs = construction.removeLifeCycleCosts()
      costs_removed = bool(removed_costs)

    removed_costs = construction.removeLifeCycleCosts()
    costs_removed = bool(removed_costs)

    # add lifeCycleCost objects if there is a non-zero value in one of the cost arguments
    if costs_requested == True:

      ureg = UnitRegistry()
      # converting doubles to si values from ip
      material_cost_si = (material_cost_ip * ureg("1/ft^2")).to("1/m^2").magnitude
      demolition_cost_si = (demolition_cost_ip * ureg("1/ft^2")).to("1/m^2").magnitude
      om_cost_si = (om_cost_ip * ureg("1/ft^2")).to("1/m^2").magnitude

      # Adding new cost items
      lcc_mat = openstudio.model.LifeCycleCost.createLifeCycleCost(
      f"LCC_Mat-{construction.name}", construction, material_cost_si, 
      "CostPerArea", "Construction", expected_life, years_until_costs_start)

      # If demo_cost_initial_const is true, later will add one-time demo costs using removed baseline objects.
      # Cost will occur at year specified by years_until_costs_start
      lcc_demo = openstudio.model.LifeCycleCost.createLifeCycleCost(
      f"LCC_Demo-{construction.name}", construction, demolition_cost_si, 
      "CostPerArea", "Salvage", expected_life, years_until_costs_start + expected_life)

      lcc_om = openstudio.model.LifeCycleCost.createLifeCycleCost(
      f"LCC_OM-{construction.name}", construction, om_cost_si, 
      "CostPerArea", "Maintenance", om_frequency, 0)

    # loop through sub surfaces
    starting_exterior_windows_constructions = []
    sub_surfaces_to_change = []
    sub_surfaces = model.getSubSurfaces()

    for sub_surface in sub_surfaces:
      if (sub_surface.outsideBoundaryCondition() == "Outdoors") & (sub_surface.subSurfaceType() == "FixedWindow") & (change_fixed_windows == True):
        sub_surfaces_to_change.append(sub_surface)
        sub_surface_const = sub_surface.construction()
        if sub_surface_const != None and sub_surface_const:
          if starting_exterior_windows_constructions:
            starting_exterior_windows_constructions.append(sub_surface_const.get().nameString())
          else:
            starting_exterior_windows_constructions.append(sub_surface_const.get().nameString())

      elif (sub_surface.outsideBoundaryCondition() == 'Outdoors') & (sub_surface.subSurfaceType() == 'OperableWindow') & (change_operable_windows == True):
        sub_surfaces_to_change.append(sub_surface)
        sub_surface_const = sub_surface.construction()
        if sub_surface_const != None and sub_surface_const:
          if starting_exterior_windows_constructions:
            starting_exterior_windows_constructions.append(sub_surface_const.get().nameString())
          else:
            starting_exterior_windows_constructions.append(sub_surface_const.get().nameString())

    # create array of constructions for sub_surfaces to change, before construction is replaced
    constructions_to_change = []
    for sub_surface in sub_surfaces_to_change:
      sub_surface_const = sub_surface.construction()
      if sub_surface_const != None and sub_surface_const:
        constructions_to_change.append(sub_surface_const.get())

    # getting cost of all existing windows before constructions are swapped. This will create demo cost if all windows were removed. Will adjust later for windows left in place
    for construction_to_change in constructions_to_change:
      # loop through lifecycle costs getting total costs under "Salvage" category
      demo_LCCs = construction_to_change.lifeCycleCosts()
      for demo_LCC in demo_LCCs:
        if demo_LCC.category() == 'Salvage':
          demo_costs_of_baseline_objects += demo_LCC.totalCost()

    if (change_fixed_windows == False) & (change_operable_windows == False):
      runner.registerAsNotApplicable("Fixed and operable windows are both set not to change.")
      return True # no need to waste time with the measure if we know it isn't applicable
    elif sub_surfaces_to_change:
      runner.registerAsNotApplicable("There are no appropriate exterior windows to change in the model.")
      return True # no need to waste time with the measure if we know it isn't applicable

    # report initial condition
    # Report initial condition
    runner.registerInitialCondition(
      f"The building had {len(set(starting_exterior_windows_constructions))} window constructions: "
      f"{', '.join(sorted(set(starting_exterior_windows_constructions)))}.")

    # loop through construction sets used in the model
    default_construction_sets = model.getDefaultConstructionSets()
    for default_construction_set in default_construction_sets:
      if default_construction_set.directUseCount() > 0:
        default_sub_surface_const_set = default_construction_set.defaultExteriorSubSurfaceConstructions()
        if default_sub_surface_const_set != None and default_sub_surface_const_set:
          starting_construction = default_sub_surface_const_set.get().fixedWindowConstruction()

          # creating new default construction set
          new_default_construction_set = default_construction_set.clone(model)
          new_default_construction_set = new_default_construction_set.to_DefaultConstructionSet().get()

          # create new sub_surface set
          new_default_sub_surface_const_set = default_sub_surface_const_set.get.clone(model)
          new_default_sub_surface_const_set = new_default_sub_surface_const_set.to_DefaultSubSurfaceConstructions().get()

          if change_fixed_windows == True:
            # assign selected construction sub_surface set
            new_default_sub_surface_const_set.setFixedWindowConstruction(construction)

          if change_operable_windows == True:
            # assign selected construction sub_surface set
            new_default_sub_surface_const_set.setOperableWindowConstruction(construction)

          # link new subset to new set
          new_default_construction_set.setDefaultExteriorSubSurfaceConstructions(new_default_sub_surface_const_set)

          # swap all uses of the old construction set for the new
          construction_set_sources = default_construction_set.sources()
          for construction_set_source in construction_set_sources:
            building_source = construction_set_source.to_Building()
            if building_source != None and building_source:
              building_source = building_source.get()
              building_source.setDefaultConstructionSet(new_default_construction_set)
              continue
            # add SpaceType, BuildingStory, and Space if statements

    # loop through appropriate sub surfaces and change where there is a hard assigned construction
    for sub_surface in sub_surfaces_to_change:
      if sub_surface.isConstructionDefaulted() != None and sub_surface.isConstructionDefaulted():
        sub_surface.setConstruction(construction)

    # loop through lifecycle costs getting total costs under "Salvage" category
    for construction_to_change in constructions_to_change:
      demo_LCCs = construction_to_change.lifeCycleCosts
      for demo_LCC in demo_LCCs:
        if demo_LCC.category == "Salvage":
          demo_costs_of_baseline_objects += demo_LCC.totalCost * -1 # this is to adjust demo cost down for original windows that were not changed

    # loop through lifecycle costs getting total costs under "Construction" or "Salvage" category and add to counter if occurs during year 0
    const_LCCs = construction.lifeCycleCosts()
    yr0_capital_totalCosts = 0
    for const_LCC in const_LCCs:
      if const_LCC.category() == "Construction" or const_LCC.category() == "Salvage":
        if const_LCC.yearsFromStart == 0:
          yr0_capital_totalCosts += const_LCC.totalCost()

    # add one time demo cost of removed windows if appropriate
    if demo_cost_initial_const == True:
      building = model.getBuilding()
      lcc_baseline_demo = openstudio.model.LifeCycleCost.createLifeCycleCost(
      'LCC_baseline_demo', 
      building, 
      demo_costs_of_baseline_objects, 
      'CostPerEach', 
      'Salvage', 
      0, 
      years_until_costs_start
      ).get()  # Using 0 for repeat period since it's a one-time cost.

      runner.registerInfo(
      f"Adding one time cost of ${neat_numbers(lcc_baseline_demo.totalCost, 0)} related to demolition of baseline objects.")

      # if demo occurs on year 0 then add to initial capital cost counter
      if lcc_baseline_demo.yearsFromStart == 0:
        yr0_capital_totalCosts += lcc_baseline_demo.totalCost()

    # ip construction area for reporting
    const_area_ip = openstudio.convert(
      openstudio.Quantity(construction.getNetArea(), openstudio.createUnit('m^2').get()),
      openstudio.createUnit('ft^2').get()).get().value()
    # get names from constructions to change
    const_names = []
    if constructions_to_change != None and constructions_to_change:
      for const_name in sorted(set(constructions_to_change)):
        const_names.append(const_name.name)

    # need to format better. At first I did each do, but seems initial condition only reports the first one.
    runner.registerFinalCondition(
      f"{neat_numbers(const_area_ip, 0)} (ft²) of existing windows of the types: {', '.join(const_names)} were replaced by new {construction.name} windows."
      f"Initial capital costs associated with the new windows are ${neat_numbers(yr0_capital_totalCosts, 0)}.")

    return True

# this allows the measure to be used by the application
ReplaceExteriorWindowConstruction().registerWithApplication()
